// postgresql-client/dist/Pool.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pool = void 0;
const lightning_pool_1 = require("lightning-pool");
const putil_varhelpers_1 = require("putil-varhelpers");
const definitions_1 = require("./definitions");
const connection_config_1 = require("./util/connection-config");
const SafeEventEmitter_1 = require("./SafeEventEmitter");
const Connection_1 = require("./Connection");
const IntlConnection_1 = require("./IntlConnection");
const common_1 = require("./common");
const debug_1 = __importDefault(require("debug"));
const debug = (0, debug_1.default)('pgc:connection');
class Pool extends SafeEventEmitter_1.SafeEventEmitter {
    constructor(config) {
        super();
        const cfg = (0, connection_config_1.getConnectionConfig)(config);
        this.config = Object.freeze(cfg);
        const poolOptions = {};
        poolOptions.acquireMaxRetries = (0, putil_varhelpers_1.coerceToInt)(cfg.acquireMaxRetries, 0);
        poolOptions.acquireRetryWait = (0, putil_varhelpers_1.coerceToInt)(cfg.acquireRetryWait, 2000);
        poolOptions.acquireTimeoutMillis = (0, putil_varhelpers_1.coerceToInt)(cfg.acquireTimeoutMillis, 0);
        poolOptions.idleTimeoutMillis = (0, putil_varhelpers_1.coerceToInt)(cfg.idleTimeoutMillis, 30000);
        poolOptions.max = (0, putil_varhelpers_1.coerceToInt)(cfg.max, 10);
        poolOptions.maxQueue = (0, putil_varhelpers_1.coerceToInt)(cfg.maxQueue, 1000);
        poolOptions.max = (0, putil_varhelpers_1.coerceToInt)(cfg.max, 10);
        poolOptions.min = (0, putil_varhelpers_1.coerceToInt)(cfg.min, 0);
        poolOptions.minIdle = (0, putil_varhelpers_1.coerceToInt)(cfg.minIdle, 0);
        poolOptions.validation = (0, putil_varhelpers_1.coerceToBoolean)(cfg.validation, false);
        const poolFactory = {
            create: async () => {
                const intlCon = new IntlConnection_1.IntlConnection(cfg);
                await intlCon.connect();
                intlCon.on('close', () => this._pool.destroy(intlCon));
                debug('created connection %s', intlCon.processID);
                return intlCon;
            },
            destroy: intlCon => {
                debug('destroy connection %s', intlCon.processID);
                return intlCon.close();
            },
            reset: async (intlCon) => {
                debug('reset connection %s', intlCon.processID);
                try {
                    if (intlCon.state === definitions_1.ConnectionState.READY)
                        await intlCon.execute('ROLLBACK;');
                }
                finally {
                    intlCon.removeAllListeners();
                    intlCon.once('close', () => this._pool.destroy(intlCon));
                    intlCon._refCount = 0;
                }
            },
            validate: async (intlCon) => {
                debug('validate connection %s', intlCon.processID);
                if (intlCon.state !== definitions_1.ConnectionState.READY)
                    throw new Error('Connection is not active');
                await intlCon.execute('select 1;');
            },
        };
        this._pool = new lightning_pool_1.Pool(poolFactory, poolOptions);
        this._pool.on('return', (...args) => this.emit('release', ...args));
        this._pool.on('error', (...args) => this.emit('error', ...args));
        this._pool.on('acquire', (...args) => this.emit('acquire', ...args));
        this._pool.on('destroy', (...args) => this.emit('destroy', ...args));
        this._pool.start();
    }
    /**
     * Returns number of connections that are currently acquired
     */
    get acquiredConnections() {
        return this._pool.acquired;
    }
    /**
     * Returns number of unused connections in the pool
     */
    get idleConnections() {
        return this._pool.available;
    }
    /**
     * Returns total number of connections in the pool regardless of whether they are idle or in use
     */
    get totalConnections() {
        return this._pool.size;
    }
    /**
     * Obtains a connection from the connection pool
     */
    async acquire() {
        const intlCon = await this._pool.acquire();
        debug('acquired connection %s', intlCon.processID);
        return new Connection_1.Connection(this, intlCon);
    }
    /**
     * Shuts down the pool and destroys all resources.
     */
    async close(terminateWait) {
        const ms = terminateWait == null ? 10000 : terminateWait;
        return this._pool.closeAsync(ms);
    }
    /**
     * Executes a script
     */
    async execute(sql, options) {
        const connection = await this.acquire();
        try {
            return await connection.execute(sql, options);
        }
        finally {
            await this.release(connection);
        }
    }
    /**
     * Executes a query
     */
    async query(sql, options) {
        const connection = await this.acquire();
        try {
            return await connection.query(sql, options);
        }
        finally {
            await this.release(connection);
        }
    }
    async prepare(sql, options) {
        debug('prepare | %s', sql);
        const connection = await this.acquire();
        const statement = await connection.prepare(sql, options);
        statement.once('close', () => this._pool.release((0, common_1.getIntlConnection)(connection)));
        return statement;
    }
    release(connection) {
        return this._pool.releaseAsync((0, common_1.getIntlConnection)(connection));
    }
}
exports.Pool = Pool;
